composeApp/src/androidMain/kotlin/io/github/couchtracker/db/app/UserManager.kt

```diff
@@ -0,0 +1,155 @@
+    ) {
+        suspend fun reloadFullUserData(coroutineContext: CoroutineContext = Dispatchers.IO) {
+            if (fullUserDataState.value !is FullUserDataState.Loaded) {
+                fullUserDataState.value = FullUserDataState.Loading
+            }
+            val result = coroutineScope.async(coroutineContext) {
+                db.read { userData ->
+                    FullUserData.load(db = userData)
+                }
+            }.await()
+            fullUserDataState.value = when (result) {
+                is UserDbResult.Completed.Success -> FullUserDataState.Loaded(result.result)
+                is UserDbResult.AnyError -> FullUserDataState.Error(result)
+            }
+        }
+
+        suspend fun ensureFullUserData() {
+            when (fullUserDataState.value) {
+                is FullUserDataState.Error, FullUserDataState.Idle -> reloadFullUserData()
+                is FullUserDataState.Loaded, FullUserDataState.Loading -> {}
+            }
+        }
+
+        suspend fun <T> write(coroutineContext: CoroutineContext = Dispatchers.IO, block: DatabaseTransaction<T>): UserDbResult<T> {
+            return coroutineScope.async(coroutineContext) {
+                db.write(block).also {
+                    if (it is UserDbResult.Completed.Success) {
+                        reloadFullUserData()
+                    }
+                }
+            }.await()
+        }
+
+        suspend fun delete(appDb: AppData, coroutineContext: CoroutineContext = Dispatchers.IO) {
+            coroutineScope.launch(coroutineContext) {
+                db.unlink()
+                appDb.userQueries.delete(user.id)
+            }.join()
+        }
+    }
+
+    private val usersInfoState = mutableStateOf(
+        UsersInfo(
+            users = initialUsers.associate {
+                it.id to UserInfo(user = it, db = it.db(), fullUserDataState = mutableStateOf(FullUserDataState.Idle))
+            },
+            currentUserId = initialUsers.first().id,
+        ),
+    )
+    private var usersInfo by usersInfoState
+
+    val users get() = usersInfo.users.map { it.value }
+    val current get() = usersInfo.currentUserInfo
+
+    init {
+        // Ensure data starts loading
+        update(usersInfo)
+    }
+
+    // TODO shouldn't be exposed?
+    fun updateListOfUsers(users: List<User>) {
+        require(users.isNotEmpty())
+        update(usersInfo.computeNew(newUsers = users))
+    }
+
+    fun changeLoggedUser(user: User) {
+        update(usersInfo.computeNew(newCurrentUserId = user.id))
+    }
+
+    private fun update(newUsersInfo: UsersInfo) {
+        val changedUser = usersInfo.currentUserId != newUsersInfo.currentUserId
+        if (changedUser) {
+            usersInfo.currentUserInfo.fullUserDataState.value = FullUserDataState.Idle
+        }
+        usersInfo = newUsersInfo
+        coroutineScope.launch {
+            newUsersInfo.currentUserInfo.ensureFullUserData()
```

Comment:
Maybe there's a bug here: we should use the new current user, rather than what's on `newUsersInfo`, as it might change between when the coroutine launches and when it gets executed.

Reply:
I don't get this comment. We _are _using the new current user. `newUsersInfo` is what's passed to the function and is an immutable class.

Reply:
If the current user changes before the job spawned here starts, we'll end up loading data for an user that's not logged in.
E.g. if someone launches two coroutines, one that does `changeLoggedUser(A)`, and one that does `changeLoggedUser(B)`, then data for both A and B will be loaded.